server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamMetadata.java

```diff
@@ -0,0 +1,170 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.cluster.Diff;
+import org.elasticsearch.cluster.DiffableUtils;
+import org.elasticsearch.cluster.NamedDiff;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Custom {@link MetaData} implementation for storing a map of {@link DataStream}s and their names.
+ */
+public class DataStreamMetadata implements MetaData.Custom {
+
+    public static final String TYPE = "data_stream";
+    private static final ParseField DATA_STREAM = new ParseField("data_stream");
+    @SuppressWarnings("unchecked")
+    private static final ConstructingObjectParser<DataStreamMetadata, Void> PARSER = new ConstructingObjectParser<>(TYPE, false,
+        a -> new DataStreamMetadata((Map<String, DataStream>) a[0]));
+
+    static {
+        PARSER.declareObject(ConstructingObjectParser.constructorArg(), (p, c) -> {
+            Map<String, DataStream> dataStreams = new HashMap<>();
+            while (p.nextToken() != XContentParser.Token.END_OBJECT) {
+                String name = p.currentName();
+                dataStreams.put(name, DataStream.fromXContent(p));
+            }
+            return dataStreams;
+        }, DATA_STREAM);
+    }
+
+    private final Map<String, DataStream> dataStreams;
+
+    public DataStreamMetadata(Map<String, DataStream> dataStreams) {
+        this.dataStreams = dataStreams;
+    }
+
+    public DataStreamMetadata(StreamInput in) throws IOException {
+        this.dataStreams = in.readMap(StreamInput::readString, DataStream::new);
+    }
+
+    public Map<String, DataStream> dataStreams() {
+        return this.dataStreams;
+    }
+
+    @Override
+    public Diff<MetaData.Custom> diff(MetaData.Custom before) {
+        return new DataStreamMetadata.DataStreamMetadataDiff((DataStreamMetadata) before, this);
+    }
+
+    public static NamedDiff<MetaData.Custom> readDiffFrom(StreamInput in) throws IOException {
+        return new DataStreamMetadata.DataStreamMetadataDiff(in);
+    }
+
+    @Override
+    public EnumSet<MetaData.XContentContext> context() {
+        return MetaData.ALL_CONTEXTS;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return TYPE;
+    }
+
+    @Override
+    public Version getMinimalSupportedVersion() {
+        return Version.V_7_7_0;
```

Comment:
What's the correct version to put here?

Reply:
When it is back ported this is the correct version, until then V_8_0_0 should be used.